File: src/test/java/pcb/uwu/core/UnitCounterTest.java
package pcb.uwu.core;

import org.junit.Test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;
import static pcb.uwu.units.fundamental.ElectricCurrentUnit.AMPERE;
import static pcb.uwu.units.fundamental.MassUnit.GRAM;
import static pcb.uwu.units.fundamental.TimeUnit.SECOND;

public class UnitCounterTest {
	@Test
	public void testObjectCounter() {
		UnitCounter test = new UnitCounter();

		assertEquals(0, test.get(SECOND));
		assertEquals(1, (test = test.major(SECOND)).get(SECOND));
		assertEquals(0, (test = test.minor(SECOND)).get(SECOND));
		assertEquals(-1, (test = test.minor(SECOND)).get(SECOND));
		assertEquals(0, (test = test.major(SECOND)).get(SECOND));
		assertEquals(1, (test = test.major(SECOND)).get(SECOND));
		assertEquals(1, (test = test.major(AMPERE)).get(AMPERE));
		assertEquals(2, (test = test.major(SECOND)).get(SECOND));
		assertEquals(3, (test = test.major(SECOND)).get(SECOND));
		assertEquals(-1, (test = test.minor(GRAM)).get(GRAM));
		assertEquals(1, test.getNegativeKeys().size());
		assertTrue(test.getNegativeKeys().contains(GRAM));
		assertEquals(2, test.getPositiveKeys().size());
		assertTrue(test.getPositiveKeys().contains(SECOND));
		assertTrue(test.getPositiveKeys().contains(AMPERE));
		assertEquals(0, (test = test.major(GRAM)).get(GRAM));
		assertEquals(0, test.getNegativeKeys().size());
		assertEquals(2, test.getPositiveKeys().size());
		assertTrue(test.getPositiveKeys().contains(SECOND));
		assertTrue(test.getPositiveKeys().contains(AMPERE));
	}
}

File: dimension/src/test/java/pcb/uwu/core/UnitCounterTest.kt
package pcb.uwu.core

import pcb.uwu.unit.base.ElectricCurrentUnit
import pcb.uwu.unit.base.MassUnit
import pcb.uwu.unit.base.TimeUnit
import pcb.uwu.util.identity
import kotlin.test.Test
import kotlin.test.assertEquals


class UnitCounterTest
{
    @Test
    fun testObjectCounter()
    {
        var test = UnitCounter()
        assertEquals(0, test[SECOND].count.toLong())
        assertEquals(1, test.major(SECOND).also { test = it }[SECOND].count)
        assertEquals(0, test.minor(SECOND).also { test = it }[SECOND].count)
        assertEquals(-1, test.minor(SECOND).also { test = it }[SECOND].count)
        assertEquals(0, test.major(SECOND).also { test = it }[SECOND].count)
        assertEquals(1, test.major(SECOND).also { test = it }[SECOND].count)
        assertEquals(1, test.major(AMPERE).also { test = it }[AMPERE].count)
        assertEquals(2, test.major(SECOND).also { test = it }[SECOND].count)
        assertEquals(3, test.major(SECOND).also { test = it }[SECOND].count)
        assertEquals(-1, test.minor(KILOGRAM).also { test = it }[KILOGRAM].count)
        assertEquals(3, test.baseUnits.size)
        assertEquals(-1, test[KILOGRAM].count)
        assertEquals(3, test[SECOND].count)
        assertEquals(1, test[AMPERE].count)
        assertEquals(0, test.major(KILOGRAM).also { test = it }[KILOGRAM].count)
        assertEquals(2, test.baseUnits.size)
        assertEquals(3, test[SECOND].count)
        assertEquals(1, test[AMPERE].count)
    }

    companion object
    {
        val SECOND: BaseUnit = TimeUnit("s", "", "", identity(), identity())
        val AMPERE: BaseUnit = ElectricCurrentUnit("A", "", "", identity(), identity())
        val KILOGRAM: BaseUnit = MassUnit("Kg", "", "", identity(), identity())
    }
}
